src/formula_engine.rs

```diff
@@ -0,0 +1,53 @@
+// License: MIT
+// Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+use std::fmt::Debug;
+use std::{
+    ops::{Add, Div, Mul, Neg, Sub},
+    str::FromStr,
+};
+
+use pest::{iterators::Pairs, Parser};
+
+use crate::{
+    error::FormulaError,
+    expression::Expr,
+    parser::{FormulaParser, Rule},
+};
+
+/// FormulaEngine holds the parsed expression and can calculate the result
+/// based on the provided component values.
+#[derive(Debug)]
+pub struct FormulaEngine<T> {
+    expr: Expr<T>,
+}
+
+impl<
+        'a,
+        T: FromStr
+            + Copy
+            + Neg<Output = T>
+            + Add<Output = T>
+            + Sub<Output = T>
+            + Mul<Output = T>
+            + Div<Output = T>
+            + PartialOrd,
+    > FormulaEngine<T>
+where
+    Expr<T>: TryFrom<Pairs<'a, Rule>>,
+    <T as FromStr>::Err: Debug,
+    FormulaError: From<<Expr<T> as TryFrom<Pairs<'a, Rule>>>::Error>,
+{
+    /// Create a new FormulaEngine from a formula string.
+    pub fn try_new(s: &'a str) -> Result<Self, FormulaError> {
+        let pairs = FormulaParser::parse(Rule::formula, s)?;
+        let expr = Expr::try_from(pairs)?;
+
+        Ok(Self { expr })
+    }
+
+    /// Calculate the result of the formula based on the provided component values.
+    pub fn calculate(&self, values: &[Option<T>]) -> Result<Option<T>, FormulaError> {
```

Comment:
We need to take a `HashMap<usize, Option<T>` or something similar here because the outside world doesn't know the order in which the values need to be.

Reply:
Yes, makes sense. Done!